costmap_3d node dies by bad_alloc if map_overlay is outside of a global map

When the `costmap_3d` node subscribes `/map_overly` topic which is completely out of  a global `/map`, this node dies by ` std::bad_alloc`.

This is because that internally negative value can be assigned to width or height of CSpace3DUpdate in this situation. ([code](https://github.com/at-wat/neonavigation/blob/8592033b44f61c3f0cf216998f23c90659379c36/costmap_cspace/include/costmap_cspace/costmap_3d_layer/output.h#L122-L123))
As a result, message that has too large (nealy max of uint32) size of data array in `CSpace3DUpdate` will be published and finally the node dies.

It would be better to ignore publishing `/costmap_update` topic  if `/map_overly` topic is completely out of the global map.

<details>
<summary>log</summary>

```
$ roslaunch neonavigation_launch demo.launch 
... logging to /home/ykoga/.ros/log/d1a3381e-a6f2-11e9-834d-186590dd30e5/roslaunch-ykoga-20734.log
Checking log directory for disk usage. This may take awhile.
Press Ctrl-C to interrupt
Done checking log file disk usage. Usage is <1GB.

started roslaunch server http://ykoga:41849/

SUMMARY
========

PARAMETERS
 * /costmap_3d/ang_resolution: 16
 * /costmap_3d/footprint: [[0.35, -0.22], [...
 * /costmap_3d/layers: [{'overlay_mode':...
 * /costmap_3d/linear_expand: 0.08
 * /costmap_3d/linear_spread: 0.3
 * /costmap_3d/static_layers: [{'type': 'Costma...
 * /largemap_to_map/hz: 1.5
 * /largemap_to_map/width: 60
 * /neonavigation_compatible: 1
 * /planner_3d/freq: 2.0
 * /planner_3d/freq_min: 0.5
 * /planner_3d/local_range: 2.0
 * /planner_3d/num_threads: 2
 * /planner_3d/use_path_with_velocity: True
 * /rosdistro: kinetic
 * /rosversion: 1.12.14
 * /spur/curv_forward: 0.2
 * /spur/dist_lim: 0.5
 * /spur/goal_tolerance_ang: 0.05
 * /spur/goal_tolerance_dist: 0.05
 * /spur/hz: 30.0
 * /spur/k_ang: 3.0
 * /spur/k_avel: 4.0
 * /spur/k_dist: 4.5
 * /spur/limit_vel_by_avel: True
 * /spur/look_forward: 0.1
 * /spur/max_acc: 0.3
 * /spur/max_angacc: 1.0
 * /spur/max_angvel: 0.6
 * /spur/max_vel: 0.5
 * /spur/no_position_control_dist: 0.05
 * /spur/path_step: 1
 * /spur/rotate_ang: 0.2
 * /spur/stop_tolerance_ang: 0.04
 * /spur/stop_tolerance_dist: 0.04

NODES
  /
    costmap_3d (costmap_cspace/costmap_3d)
    dummy_robot (planner_cspace/dummy_robot)
    largemap_to_map (costmap_cspace/largemap_to_map)
    map_server (map_server/map_server)
    map_server_local (map_server/map_server)
    patrol (planner_cspace/patrol)
    planner_3d (planner_cspace/planner_3d)
    rviz (rviz/rviz)
    spur (trajectory_tracker/trajectory_tracker)
    stf1 (tf2_ros/static_transform_publisher)

auto-starting new master
process[master]: started with pid [20744]
ROS_MASTER_URI=http://localhost:11311

setting /run_id to d1a3381e-a6f2-11e9-834d-186590dd30e5
process[rosout-1]: started with pid [20757]
started core service [/rosout]
process[costmap_3d-2]: started with pid [20774]
process[map_server-3]: started with pid [20775]
process[planner_3d-4]: started with pid [20789]
process[spur-5]: started with pid [20797]
process[patrol-6]: started with pid [20808]
process[dummy_robot-7]: started with pid [20828]
process[stf1-8]: started with pid [20838]
process[map_server_local-9]: started with pid [20847]
process[largemap_to_map-10]: started with pid [20860]
[ INFO] [1563189746.615713931]: New static layer: unknown
[ WARN] [1563189746.615801620]: overlay_mode of the static layer is not specified. Using MAX mode.
process[rviz-11]: started with pid [20879]
[ INFO] [1563189746.674812552]: New static layer: overlay0
[ INFO] [1563189746.814571052]: New layer: overlay2
[ INFO] [1563189746.866349534]: New layer: overlay1
[ INFO] [1563189746.956683713]: 2D costmap received
[ INFO] [1563189748.707477684]: Map received
[ INFO] [1563189748.707561722]:  linear_resolution 0.05 x (400x400) px
[ INFO] [1563189748.707596183]:  angular_resolution 0.39 x 16 px
[ INFO] [1563189748.707625633]:  origin -10.000 m, -10.000 m, 0.000 rad
[ INFO] [1563189749.258053933]: x:7, y:7 grids around the boundary is ignored on path search
[ INFO] [1563189773.705720901]: New goal received (264, 293, 4)
[ERROR] [1563189778.758714546]: Position jumped (12.282/1.000, 0.245/1.500); clearing history
[ WARN] [1563189778.758859055]: Oops! You are in Rock!
[ERROR] [1563189779.209776987]: You are on the edge of the world.
[ WARN] [1563189779.209847529]: Oops! You are in Rock!
[ WARN] [1563189779.705791926]: Oops! You are in Rock!
[ WARN] [1563189780.214125122]: Oops! You are in Rock!
[ WARN] [1563189780.705717191]: Oops! You are in Rock!
[ WARN] [1563189781.222154753]: Oops! You are in Rock!
[ WARN] [1563189781.705698776]: Oops! You are in Rock!
[ WARN] [1563189782.205709201]: Oops! You are in Rock!
[ WARN] [1563189782.710184015]: Oops! You are in Rock!
[ WARN] [1563189783.205734355]: Oops! You are in Rock!
[ WARN] [1563189783.718114774]: Oops! You are in Rock!
[ WARN] [1563189784.205734110]: Oops! You are in Rock!
[ WARN] [1563189784.726240091]: Oops! You are in Rock!
[ WARN] [1563189785.206115052]: Oops! You are in Rock!
[ WARN] [1563189785.705756589]: Oops! You are in Rock!
[ WARN] [1563189786.214203820]: Oops! You are in Rock!
[ WARN] [1563189786.705734343]: Oops! You are in Rock!
[ WARN] [1563189787.222203333]: Oops! You are in Rock!
[ WARN] [1563189787.705742892]: Oops! You are in Rock!
[ WARN] [1563189788.205723615]: Oops! You are in Rock!
[ WARN] [1563189788.710222021]: Oops! You are in Rock!
[ WARN] [1563189789.205718794]: Oops! You are in Rock!
[ WARN] [1563189789.718224281]: Oops! You are in Rock!
[ERROR] [1563189789.884400708]: a message of over a gigabyte was predicted in tcpros. that seems highly unlikely, so I'll assume protocol synchronization is lost.
[ WARN] [1563189790.205727747]: Oops! You are in Rock!
[ WARN] [1563189790.726136511]: Oops! You are in Rock!
[ WARN] [1563189791.206243766]: Oops! You are in Rock!
[ WARN] [1563189791.705721752]: Oops! You are in Rock!
[ WARN] [1563189792.214127410]: Oops! You are in Rock!
[ WARN] [1563189792.705723930]: Oops! You are in Rock!
[ WARN] [1563189793.222183517]: Oops! You are in Rock!
[ WARN] [1563189793.705767665]: Oops! You are in Rock!
[ WARN] [1563189794.205719503]: Oops! You are in Rock!
[ WARN] [1563189794.710166384]: Oops! You are in Rock!
terminate called after throwing an instance of 'std::bad_alloc'
  what():  std::bad_alloc
[ WARN] [1563189795.205720843]: Oops! You are in Rock!
[ WARN] [1563189795.718130766]: Oops! You are in Rock!
[costmap_3d-2] process has died [pid 20774, exit code -6, cmd /home/ykoga/sq_ws/devel/lib/costmap_cspace/costmap_3d __name:=costmap_3d __log:=/home/ykoga/.ros/log/d1a3381e-a6f2-11e9-834d-186590dd30e5/costmap_3d-2.log].
log file: /home/ykoga/.ros/log/d1a3381e-a6f2-11e9-834d-186590dd30e5/costmap_3d-2*.log
[ WARN] [1563189796.205688303]: Oops! You are in Rock!
[ WARN] [1563189796.705858076]: Oops! You are in Rock!
[ WARN] [1563189797.206289667]: Oops! You are in Rock!
[ WARN] [1563189797.705738850]: Oops! You are in Rock!
[ WARN] [1563189798.205808489]: Oops! You are in Rock!
^C[rviz-11] killing on exit
[largemap_to_map-10] killing on exit
[map_server_local-9] killing on exit
[stf1-8] killing on exit
[dummy_robot-7] killing on exit
[patrol-6] killing on exit
[spur-5] killing on exit
[planner_3d-4] killing on exit
[map_server-3] killing on exit
[rosout-1] killing on exit
[master] killing on exit
shutting down processing monitor...
... shutting down processing monitor complete
done
neonavigation (master $% u+16) 
```

</details>

![visualization rviz- - RViz_480](https://user-images.githubusercontent.com/6742467/61213208-53125780-a73f-11e9-8ca9-d6c5bbccb3a1.png)

--- a/costmap_cspace/include/costmap_cspace/costmap_3d_layer/output.h
+++ b/costmap_cspace/include/costmap_cspace/costmap_3d_layer/output.h
@@ -57,7 +57,7 @@ public:
   {
     cb_ = cb;
   }
-  void setMapMetaData(const costmap_cspace_msgs::MapMetaData3D &info)
+  void setMapMetaData(const costmap_cspace_msgs::MapMetaData3D& info)
   {
   }
 
@@ -74,8 +74,8 @@ protected:
     return true;
   }
   void updateCSpace(
-      const nav_msgs::OccupancyGrid::ConstPtr &map,
-      const UpdatedRegion &region)
+      const nav_msgs::OccupancyGrid::ConstPtr& map,
+      const UpdatedRegion& region)
   {
   }
   costmap_cspace_msgs::CSpace3DUpdate::Ptr generateUpdateMsg()
@@ -148,10 +148,10 @@ protected:
           const int y2 = update_msg->y + j;
           const int yaw2 = update_msg->yaw + k;
 
-          const auto &m = map_->getCost(x2, y2, yaw2);
+          const auto& m = map_->getCost(x2, y2, yaw2);
           const size_t addr = (k * update_msg->height + j) * update_msg->width + i;
           ROS_ASSERT(addr < update_msg->data.size());
-          auto &up = update_msg->data[addr];
+          auto& up = update_msg->data[addr];
           up = m;
         }
       }

--- a/costmap_cspace/include/costmap_cspace/costmap_3d_layer/output.h
+++ b/costmap_cspace/include/costmap_cspace/costmap_3d_layer/output.h
@@ -57,7 +57,7 @@ public:
   {
     cb_ = cb;
   }
-  void setMapMetaData(const costmap_cspace_msgs::MapMetaData3D& info)
+  void setMapMetaData(const costmap_cspace_msgs::MapMetaData3D &info)
   {
   }
 
@@ -74,8 +74,8 @@ protected:
     return true;
   }
   void updateCSpace(
-      const nav_msgs::OccupancyGrid::ConstPtr& map,
-      const UpdatedRegion& region)
+      const nav_msgs::OccupancyGrid::ConstPtr &map,
+      const UpdatedRegion &region)
   {
   }
   costmap_cspace_msgs::CSpace3DUpdate::Ptr generateUpdateMsg()
@@ -105,6 +105,14 @@ protected:
     {
       update_height = map_->info.height - update_y;
     }
+    if (update_width < 0)
+    {
+      update_width = 0;
+    }
+    if (update_height < 0)
+    {
+      update_height = 0;
+    }
     UpdatedRegion region(
         update_x,
         update_y,
@@ -116,6 +124,11 @@ protected:
     region_merged.merge(region_prev_);
     region_prev_ = region;
     region_ = UpdatedRegion();
+
+    if (update_width == 0 || update_height == 0)
+    {
+      return nullptr;
+    }
 
     update_msg->x = region_merged.x_;
     update_msg->y = region_merged.y_;
@@ -135,10 +148,10 @@ protected:
           const int y2 = update_msg->y + j;
           const int yaw2 = update_msg->yaw + k;
 
-          const auto& m = map_->getCost(x2, y2, yaw2);
+          const auto &m = map_->getCost(x2, y2, yaw2);
           const size_t addr = (k * update_msg->height + j) * update_msg->width + i;
           ROS_ASSERT(addr < update_msg->data.size());
-          auto& up = update_msg->data[addr];
+          auto &up = update_msg->data[addr];
           up = m;
         }
       }

--- a/costmap_cspace/include/costmap_cspace/costmap_3d_layer/stop_propagation.h
+++ b/costmap_cspace/include/costmap_cspace/costmap_3d_layer/stop_propagation.h
@@ -48,7 +48,7 @@ public:
   void loadConfig(XmlRpc::XmlRpcValue config)
   {
   }
-  void setMapMetaData(const costmap_cspace_msgs::MapMetaData3D &info)
+  void setMapMetaData(const costmap_cspace_msgs::MapMetaData3D& info)
   {
   }
 
@@ -60,13 +60,13 @@ protected:
   bool updateChain(const bool output)
   {
     region_ = UpdatedRegion(0, 0, 0, 0, 0, 0, ros::Time(0));
-    for (auto &c : map_overlay_->data)
+    for (auto& c : map_overlay_->data)
       c = -1;
     return false;
   }
   void updateCSpace(
-      const nav_msgs::OccupancyGrid::ConstPtr &map,
-      const UpdatedRegion &region)
+      const nav_msgs::OccupancyGrid::ConstPtr& map,
+      const UpdatedRegion& region)
   {
   }
 };

--- a/costmap_cspace/include/costmap_cspace/costmap_3d_layer/stop_propagation.h
+++ b/costmap_cspace/include/costmap_cspace/costmap_3d_layer/stop_propagation.h
@@ -42,13 +42,13 @@ namespace costmap_cspace
 class Costmap3dLayerStopPropagation : public Costmap3dLayerBase
 {
 public:
-  using Ptr = std::shared_ptr<Costmap3dLayerOutput>;
+  using Ptr = std::shared_ptr<Costmap3dLayerStopPropagation>;
 
 public:
   void loadConfig(XmlRpc::XmlRpcValue config)
   {
   }
-  void setMapMetaData(const costmap_cspace_msgs::MapMetaData3D& info)
+  void setMapMetaData(const costmap_cspace_msgs::MapMetaData3D &info)
   {
   }
 
@@ -60,13 +60,13 @@ protected:
   bool updateChain(const bool output)
   {
     region_ = UpdatedRegion(0, 0, 0, 0, 0, 0, ros::Time(0));
-    for (auto& c : map_overlay_->data)
+    for (auto &c : map_overlay_->data)
       c = -1;
     return false;
   }
   void updateCSpace(
-      const nav_msgs::OccupancyGrid::ConstPtr& map,
-      const UpdatedRegion& region)
+      const nav_msgs::OccupancyGrid::ConstPtr &map,
+      const UpdatedRegion &region)
   {
   }
 };

--- a/costmap_cspace/src/costmap_3d.cpp
+++ b/costmap_cspace/src/costmap_3d.cpp
@@ -61,7 +61,7 @@ protected:
                 costmap_cspace::Costmap3dLayerBase::Ptr>> map_buffer_;
 
   void cbMap(
-      const nav_msgs::OccupancyGrid::ConstPtr& msg,
+      const nav_msgs::OccupancyGrid::ConstPtr &msg,
       const costmap_cspace::Costmap3dLayerBase::Ptr map)
   {
     if (map->getAngularGrid() <= 0)
@@ -76,14 +76,14 @@ protected:
 
     if (map_buffer_.size() > 0)
     {
-      for (auto& map : map_buffer_)
+      for (auto &map : map_buffer_)
         cbMapOverlay(map.first, map.second);
       ROS_INFO("%ld buffered costmaps processed", map_buffer_.size());
       map_buffer_.clear();
     }
   }
   void cbMapOverlay(
-      const nav_msgs::OccupancyGrid::ConstPtr& msg,
+      const nav_msgs::OccupancyGrid::ConstPtr &msg,
       const costmap_cspace::Costmap3dLayerBase::Ptr map)
   {
     ROS_DEBUG("Overlay 2D costmap received");
@@ -113,11 +113,18 @@ protected:
       const costmap_cspace::CSpace3DMsg::Ptr map,
       const costmap_cspace_msgs::CSpace3DUpdate::Ptr update)
   {
-    publishDebug(*map);
-    pub_costmap_update_.publish(*update);
+    if (update)
+    {
+      publishDebug(*map);
+      pub_costmap_update_.publish(*update);
+    }
+    else
+    {
+      ROS_WARN("Updated region of the costmap is empty. The position may be out-of-boundary, or input map is wrong.");
+    }
     return true;
   }
-  void publishDebug(const costmap_cspace_msgs::CSpace3D& map)
+  void publishDebug(const costmap_cspace_msgs::CSpace3D &map)
   {
     if (pub_debug_.getNumSubscribers() == 0)
       return;
@@ -141,7 +148,7 @@ protected:
     }
     pub_debug_.publish(pc);
   }
-  void cbPublishFootprint(const ros::TimerEvent& event, const geometry_msgs::PolygonStamped msg)
+  void cbPublishFootprint(const ros::TimerEvent &event, const geometry_msgs::PolygonStamped msg)
   {
     auto footprint = msg;
     footprint.header.stamp = ros::Time::now();
@@ -193,7 +200,7 @@ public:
     {
       footprint = costmap_cspace::Polygon(footprint_xml);
     }
-    catch (const std::exception& e)
+    catch (const std::exception &e)
     {
       ROS_FATAL("Invalid footprint");
       throw e;
@@ -372,7 +379,7 @@ public:
   }
 };
 
-int main(int argc, char* argv[])
+int main(int argc, char *argv[])
 {
   ros::init(argc, argv, "costmap_3d");
 

--- a/costmap_cspace/src/costmap_3d.cpp
+++ b/costmap_cspace/src/costmap_3d.cpp
@@ -61,7 +61,7 @@ protected:
                 costmap_cspace::Costmap3dLayerBase::Ptr>> map_buffer_;
 
   void cbMap(
-      const nav_msgs::OccupancyGrid::ConstPtr &msg,
+      const nav_msgs::OccupancyGrid::ConstPtr& msg,
       const costmap_cspace::Costmap3dLayerBase::Ptr map)
   {
     if (map->getAngularGrid() <= 0)
@@ -76,14 +76,14 @@ protected:
 
     if (map_buffer_.size() > 0)
     {
-      for (auto &map : map_buffer_)
+      for (auto& map : map_buffer_)
         cbMapOverlay(map.first, map.second);
       ROS_INFO("%ld buffered costmaps processed", map_buffer_.size());
       map_buffer_.clear();
     }
   }
   void cbMapOverlay(
-      const nav_msgs::OccupancyGrid::ConstPtr &msg,
+      const nav_msgs::OccupancyGrid::ConstPtr& msg,
       const costmap_cspace::Costmap3dLayerBase::Ptr map)
   {
     ROS_DEBUG("Overlay 2D costmap received");
@@ -124,7 +124,7 @@ protected:
     }
     return true;
   }
-  void publishDebug(const costmap_cspace_msgs::CSpace3D &map)
+  void publishDebug(const costmap_cspace_msgs::CSpace3D& map)
   {
     if (pub_debug_.getNumSubscribers() == 0)
       return;
@@ -148,7 +148,7 @@ protected:
     }
     pub_debug_.publish(pc);
   }
-  void cbPublishFootprint(const ros::TimerEvent &event, const geometry_msgs::PolygonStamped msg)
+  void cbPublishFootprint(const ros::TimerEvent& event, const geometry_msgs::PolygonStamped msg)
   {
     auto footprint = msg;
     footprint.header.stamp = ros::Time::now();
@@ -200,7 +200,7 @@ public:
     {
       footprint = costmap_cspace::Polygon(footprint_xml);
     }
-    catch (const std::exception &e)
+    catch (const std::exception& e)
     {
       ROS_FATAL("Invalid footprint");
       throw e;
@@ -379,7 +379,7 @@ public:
   }
 };
 
-int main(int argc, char *argv[])
+int main(int argc, char* argv[])
 {
   ros::init(argc, argv, "costmap_3d");
 

--- a/costmap_cspace/test/src/test_costmap_3d.cpp
+++ b/costmap_cspace/test/src/test_costmap_3d.cpp
@@ -107,7 +107,7 @@ TEST(Costmap3dLayerFootprint, CSpaceTemplate)
 
   ASSERT_EQ(cm.getRangeMax(), static_cast<int>(ceilf(1.5 / 1.0)));
 
-  const costmap_cspace::CSpace3Cache &temp = cm.getTemplate();
+  const costmap_cspace::CSpace3Cache& temp = cm.getTemplate();
   // Check template size
   int x, y, a;
   int cx, cy, ca;
@@ -169,7 +169,7 @@ TEST(Costmap3dLayerPlain, CSpaceTemplate)
 
   ASSERT_EQ(cm.getRangeMax(), 0);
 
-  const costmap_cspace::CSpace3Cache &temp = cm.getTemplate();
+  const costmap_cspace::CSpace3Cache& temp = cm.getTemplate();
   // Check template size
   int x, y, a;
   int cx, cy, ca;
@@ -232,7 +232,7 @@ TEST(Costmap3dLayerFootprint, CSpaceGenerate)
   }
 
   // std::cout << "========" << std::endl;
-  for (auto &g : map->data)
+  for (auto& g : map->data)
   {
     g = 100;
   }
@@ -254,7 +254,7 @@ TEST(Costmap3dLayerFootprint, CSpaceGenerate)
   }
 
   // C shape wall in the map
-  for (auto &g : map->data)
+  for (auto& g : map->data)
   {
     g = 0;
   }
@@ -443,7 +443,7 @@ TEST(Costmap3dLayerFootprint, CSpaceOverwrite)
   costmap_cspace_msgs::CSpace3DUpdate::Ptr updated(new costmap_cspace_msgs::CSpace3DUpdate);
   auto cb = [&updated](
       const costmap_cspace::CSpace3DMsg::Ptr map,
-      const costmap_cspace_msgs::CSpace3DUpdate::Ptr &update) -> bool
+      const costmap_cspace_msgs::CSpace3DUpdate::Ptr& update) -> bool
   {
     updated = update;
     return true;
@@ -498,7 +498,7 @@ TEST(Costmap3dLayerFootprint, CSpaceOverwrite)
   costmap_cspace_msgs::CSpace3DUpdate::Ptr updated_max(new costmap_cspace_msgs::CSpace3DUpdate);
   auto cb_max = [&updated_max](
       const costmap_cspace::CSpace3DMsg::Ptr map,
-      const costmap_cspace_msgs::CSpace3DUpdate::Ptr &update) -> bool
+      const costmap_cspace_msgs::CSpace3DUpdate::Ptr& update) -> bool
   {
     updated_max = update;
     return true;
@@ -683,7 +683,7 @@ TEST(Costmap3dLayerOutput, CSpaceOutOfBoundary)
         { "outside xy3", { -3.0, 5.0 }, false },
       };
 
-  for (auto &d : dataset)
+  for (auto& d : dataset)
   {
     const std::string test_name = "Case [" + d.name + "]";
     // Settings: 4 angular grids
@@ -717,8 +717,8 @@ TEST(Costmap3dLayerOutput, CSpaceOutOfBoundary)
     // Overlay local map
     costmap_cspace_msgs::CSpace3DUpdate::Ptr updated;
     auto cb = [&updated](
-        const costmap_cspace::CSpace3DMsg::Ptr &map,
-        const costmap_cspace_msgs::CSpace3DUpdate::Ptr &update) -> bool
+        const costmap_cspace::CSpace3DMsg::Ptr& map,
+        const costmap_cspace_msgs::CSpace3DUpdate::Ptr& update) -> bool
     {
       updated = update;
       return true;
@@ -743,7 +743,7 @@ TEST(Costmap3dLayerOutput, CSpaceOutOfBoundary)
   }
 }
 
-int main(int argc, char **argv)
+int main(int argc, char** argv)
 {
   testing::InitGoogleTest(&argc, argv);
 

--- a/costmap_cspace/test/src/test_costmap_3d.cpp
+++ b/costmap_cspace/test/src/test_costmap_3d.cpp
@@ -107,7 +107,7 @@ TEST(Costmap3dLayerFootprint, CSpaceTemplate)
 
   ASSERT_EQ(cm.getRangeMax(), static_cast<int>(ceilf(1.5 / 1.0)));
 
-  const costmap_cspace::CSpace3Cache& temp = cm.getTemplate();
+  const costmap_cspace::CSpace3Cache &temp = cm.getTemplate();
   // Check template size
   int x, y, a;
   int cx, cy, ca;
@@ -169,7 +169,7 @@ TEST(Costmap3dLayerPlain, CSpaceTemplate)
 
   ASSERT_EQ(cm.getRangeMax(), 0);
 
-  const costmap_cspace::CSpace3Cache& temp = cm.getTemplate();
+  const costmap_cspace::CSpace3Cache &temp = cm.getTemplate();
   // Check template size
   int x, y, a;
   int cx, cy, ca;
@@ -232,7 +232,7 @@ TEST(Costmap3dLayerFootprint, CSpaceGenerate)
   }
 
   // std::cout << "========" << std::endl;
-  for (auto& g : map->data)
+  for (auto &g : map->data)
   {
     g = 100;
   }
@@ -254,7 +254,7 @@ TEST(Costmap3dLayerFootprint, CSpaceGenerate)
   }
 
   // C shape wall in the map
-  for (auto& g : map->data)
+  for (auto &g : map->data)
   {
     g = 0;
   }
@@ -443,7 +443,7 @@ TEST(Costmap3dLayerFootprint, CSpaceOverwrite)
   costmap_cspace_msgs::CSpace3DUpdate::Ptr updated(new costmap_cspace_msgs::CSpace3DUpdate);
   auto cb = [&updated](
       const costmap_cspace::CSpace3DMsg::Ptr map,
-      const costmap_cspace_msgs::CSpace3DUpdate::Ptr& update) -> bool
+      const costmap_cspace_msgs::CSpace3DUpdate::Ptr &update) -> bool
   {
     updated = update;
     return true;
@@ -498,7 +498,7 @@ TEST(Costmap3dLayerFootprint, CSpaceOverwrite)
   costmap_cspace_msgs::CSpace3DUpdate::Ptr updated_max(new costmap_cspace_msgs::CSpace3DUpdate);
   auto cb_max = [&updated_max](
       const costmap_cspace::CSpace3DMsg::Ptr map,
-      const costmap_cspace_msgs::CSpace3DUpdate::Ptr& update) -> bool
+      const costmap_cspace_msgs::CSpace3DUpdate::Ptr &update) -> bool
   {
     updated_max = update;
     return true;
@@ -630,7 +630,126 @@ TEST(Costmap3dLayerFootprint, CSpaceOverlayMove)
   }
 }
 
-int main(int argc, char** argv)
+TEST(Costmap3dLayerOutput, CSpaceOutOfBoundary)
+{
+  struct TestData
+  {
+    struct Input
+    {
+      float x;
+      float y;
+    };
+    struct Expected
+    {
+      unsigned int x;
+      unsigned int y;
+      unsigned int yaw;
+      unsigned int width;
+      unsigned int height;
+      unsigned int angle;
+    };
+    std::string name;
+    Input input;
+    bool valid;
+    Expected expected;
+  };
+  const TestData dataset[] =
+      {
+        { "inside0", { 0.0, 0.0 }, true, { 0u, 0u, 0u, 2u, 2u, 4u } },
+        { "inside1", { 1.0, 0.0 }, true, { 1u, 0u, 0u, 2u, 2u, 4u } },
+        { "half-outside x0", { -1.0, 0.0 }, true, { 0u, 0u, 0u, 1u, 2u, 4u } },
+        { "half-outside x1", { 3.0, 0.0 }, true, { 3u, 0u, 0u, 1u, 2u, 4u } },
+        { "half-outside y0", { 0.0, -1.0 }, true, { 0u, 0u, 0u, 2u, 1u, 4u } },
+        { "half-outside y1", { 0.0, 3.0 }, true, { 0u, 3u, 0u, 2u, 1u, 4u } },
+        { "half-outside xy0", { -1.0, -1.0 }, true, { 0u, 0u, 0u, 1u, 1u, 4u } },
+        { "half-outside xy1", { 3.0, -1.0 }, true, { 3u, 0u, 0u, 1u, 1u, 4u } },
+        { "half-outside xy2", { 3.0, 3.0 }, true, { 3u, 3u, 0u, 1u, 1u, 4u } },
+        { "half-outside xy3", { -1.0, 3.0 }, true, { 0u, 3u, 0u, 1u, 1u, 4u } },
+        { "boundary x0", { -2.0, 0.0 }, false },
+        { "boundary x1", { 4, 0.0 }, false },
+        { "boundary y0", { 0, -2.0 }, false },
+        { "boundary y1", { 0, 4.0 }, false },
+        { "boundary xy0", { -2.0, -2.0 }, false },
+        { "boundary xy1", { 4.0, -2.0 }, false },
+        { "boundary xy2", { 4.0, 4.0 }, false },
+        { "boundary xy3", { -2.0, 4.0 }, false },
+        { "outside x0", { -3.0, 0.0 }, false },
+        { "outside x1", { 5, 0.0 }, false },
+        { "outside y0", { 0, -3.0 }, false },
+        { "outside y1", { 0, 5.0 }, false },
+        { "outside xy0", { -3.0, -3.0 }, false },
+        { "outside xy1", { 5.0, -3.0 }, false },
+        { "outside xy2", { 5.0, 5.0 }, false },
+        { "outside xy3", { -3.0, 5.0 }, false },
+      };
+
+  for (auto &d : dataset)
+  {
+    const std::string test_name = "Case [" + d.name + "]";
+    // Settings: 4 angular grids
+    costmap_cspace::Costmap3d cms(4);
+    auto cm = cms.addRootLayer<costmap_cspace::Costmap3dLayerPlain>();
+    auto cm_stop = cms.addLayer<costmap_cspace::Costmap3dLayerStopPropagation>();
+    auto cm_over = cms.addLayer<costmap_cspace::Costmap3dLayerPlain>(
+        costmap_cspace::MapOverlayMode::OVERWRITE);
+    auto cm_output = cms.addLayer<costmap_cspace::Costmap3dLayerOutput>();
+
+    // Generate two sample maps
+    nav_msgs::OccupancyGrid::Ptr map(new nav_msgs::OccupancyGrid);
+    map->info.width = 4;
+    map->info.height = 4;
+    map->info.resolution = 1.0;
+    map->info.origin.orientation.w = 1.0;
+    map->data.resize(map->info.width * map->info.height);
+
+    nav_msgs::OccupancyGrid::Ptr map2(new nav_msgs::OccupancyGrid);
+    map2->info.width = 2;
+    map2->info.height = 2;
+    map2->info.resolution = 1.0;
+    map2->info.origin.orientation.w = 1.0;
+    map2->info.origin.position.x = d.input.x;
+    map2->info.origin.position.y = d.input.y;
+    map2->data.resize(map->info.width * map->info.height);
+
+    // Apply base map
+    cm->setBaseMap(map);
+
+    // Overlay local map
+    auto cb1 = [](
+        const costmap_cspace::CSpace3DMsg::Ptr &map,
+        const costmap_cspace_msgs::CSpace3DUpdate::Ptr &update) -> bool
+    {
+      return true;
+    };
+    costmap_cspace_msgs::CSpace3DUpdate::Ptr updated;
+    auto cb = [&updated](
+        const costmap_cspace::CSpace3DMsg::Ptr &map,
+        const costmap_cspace_msgs::CSpace3DUpdate::Ptr &update) -> bool
+    {
+      updated = update;
+      return true;
+    };
+    cm_output->setHandler(cb);
+    cm_over->processMapOverlay(map2);
+
+    if (d.valid)
+    {
+      ASSERT_TRUE(static_cast<bool>(updated)) << test_name;
+      EXPECT_EQ(updated->x, d.expected.x) << test_name;
+      EXPECT_EQ(updated->y, d.expected.y) << test_name;
+      EXPECT_EQ(updated->yaw, d.expected.yaw) << test_name;
+      EXPECT_EQ(updated->width, d.expected.width) << test_name;
+      EXPECT_EQ(updated->height, d.expected.height) << test_name;
+      EXPECT_EQ(updated->angle, d.expected.angle) << test_name;
+    }
+    else
+    {
+      EXPECT_FALSE(static_cast<bool>(updated)) << test_name;
+    }
+  }
+}
+
+int main(int argc, char **argv)
 {
   testing::InitGoogleTest(&argc, argv);
 

--- a/costmap_cspace/test/src/test_costmap_3d.cpp
+++ b/costmap_cspace/test/src/test_costmap_3d.cpp
@@ -715,12 +715,6 @@ TEST(Costmap3dLayerOutput, CSpaceOutOfBoundary)
     cm->setBaseMap(map);
 
     // Overlay local map
-    auto cb1 = [](
-        const costmap_cspace::CSpace3DMsg::Ptr &map,
-        const costmap_cspace_msgs::CSpace3DUpdate::Ptr &update) -> bool
-    {
-      return true;
-    };
     costmap_cspace_msgs::CSpace3DUpdate::Ptr updated;
     auto cb = [&updated](
         const costmap_cspace::CSpace3DMsg::Ptr &map,

